### PR TITLE
feat: RoomType

### PR DESCRIPTION
**Pull request by kristianbinau**
## Description

- Added room types

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: src/modules/room/type/__test__/browseRoomType.test.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                name: "Santa Marina Hotel",
+                description: "Santa Marina Hotel is located close to the beach",
+                address: "8130 Sv. Marina, Sozopol, Bulgarien",
+            },
+        });
+        await prisma.roomType.create({
+            data: {
+                id: 1000,
+                name: "Double room",
+                description: "Room for 2 clowns laying in one bed",
```
Comment (frederikpyt) [inline src/modules/room/type/__test__/browseRoomType.test.ts:28]: 🥇

File: src/modules/room/type/__test__/createRoomType.test.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        });
+
+        expect(response.statusCode).toBe(400);
+        expect(response.json()).toEqual({
+            error: "Bad Request",
+            message: "body/name must NOT have fewer than 1 characters",
+            statusCode: 400,
+        });
+    });
+});
```
Comment (frederikpyt) [inline src/modules/room/type/__test__/createRoomType.test.ts:139]: More tests on invalid and missing properties in payload

1. test with a hotel id that does not exist
2. test with missing price

File: src/modules/room/type/__test__/createRoomType.test.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+describe("POST /api/room/type", () => {
+    let fastify: FastifyInstance;
+
+    beforeAll(async () => {
+        fastify = await build();
+    });
+
+    beforeEach(async () => {
+        await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/room/type/__test__/createRoomType.test.ts:14]: Isn't this the wrong order as roomtypes belong to hotels?

File: src/modules/room/type/__test__/deleteRoomType.test.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+describe("DELETE /api/room/type/:id", () => {
+    let fastify: FastifyInstance;
+
+    beforeAll(async () => {
+        fastify = await build();
+    });
+
+    beforeEach(async () => {
+        await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/room/type/__test__/deleteRoomType.test.ts:14]: again

File: src/modules/hotel/__test__/browseHotel.test.ts [hunk 5]
```diff
@@ -12,8 +10,9 @@ describe("GET /api/hotel", () => {
     });
 
     beforeEach(async () => {
-        await fastify.redis.del(CACHE_KEY_HOTELS);
+        await fastify.redis.flushall();
         await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/browseHotel.test.ts:15]: again

File: src/modules/hotel/__test__/createHotel.test.ts [hunk 6]
```diff
@@ -11,6 +11,7 @@ describe("POST /api/hotel", () => {
 
     beforeEach(async () => {
         await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/createHotel.test.ts:14]: again

File: src/modules/hotel/__test__/deleteHotel.test.ts [hunk 7]
```diff
@@ -11,6 +11,7 @@ describe("DELETE /api/hotel/:id", () => {
 
     beforeEach(async () => {
         await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/deleteHotel.test.ts:14]: again

File: src/modules/hotel/__test__/showHotel.test.ts [hunk 8]
```diff
@@ -12,8 +10,9 @@ describe("GET /api/hotel/:id", () => {
     });
 
     beforeEach(async () => {
-        await fastify.redis.del(CACHE_KEY_HOTEL+"1000");
+        await fastify.redis.flushall();
         await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/showHotel.test.ts:15]: again

File: src/modules/hotel/__test__/showHotelRoomType.test.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    let fastify: FastifyInstance;
+
+    beforeAll(async () => {
+        fastify = await build();
+    });
+
+    beforeEach(async () => {
+        await fastify.redis.flushall();
+        await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/showHotelRoomType.test.ts:15]: again

File: src/modules/hotel/__test__/showHotelRoomType.test.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        await prisma.hotel.deleteMany();
+        const response = await fastify.inject({
+            method: "GET",
+            url: "/api/hotel/1002/room_types"
+        });
+
+        expect(response.statusCode).toBe(200);
+        expect(response.json()).toEqual([]);
+    });
+});
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/showHotelRoomType.test.ts:129]: Test on hotel id that doesn't exist

File: src/modules/hotel/__test__/updateHotel.test.ts [hunk 11]
```diff
@@ -11,6 +11,7 @@ describe("POST /api/hotel", () => {
 
     beforeEach(async () => {
         await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/updateHotel.test.ts:14]: again

File: src/modules/room/type/__test__/showRoomType.test.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    let fastify: FastifyInstance;
+
+    beforeAll(async () => {
+        fastify = await build();
+    });
+
+    beforeEach(async () => {
+        await fastify.redis.flushall();
+        await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/room/type/__test__/showRoomType.test.ts:15]: again

File: src/modules/room/type/__test__/updateRoomType.test.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+describe("POST /api/room/type", () => {
+    let fastify: FastifyInstance;
+
+    beforeAll(async () => {
+        fastify = await build();
+    });
+
+    beforeEach(async () => {
+        await prisma.hotel.deleteMany();
+        await prisma.roomType.deleteMany();
```
Comment (frederikpyt) [inline src/modules/room/type/__test__/updateRoomType.test.ts:14]: again

File: src/modules/room/type/__test__/updateRoomType.test.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        });
+
+        expect(response.statusCode).toBe(400);
+        expect(response.json()).toEqual({
+            error: "Bad Request",
+            message: "body/name must NOT have fewer than 1 characters",
+            statusCode: 400,
+        });
+    });
+});
```
Comment (frederikpyt) [inline src/modules/room/type/__test__/updateRoomType.test.ts:152]: More tests!

File: src/modules/room/type/type.errors.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+export async function error_message(e: unknown) {
+    let message = String(e);
+
+    if (e instanceof Error) {
+        message = e.message;
+    }
+
+    return message;
+}
```
Comment (frederikpyt) [inline src/modules/room/type/type.errors.ts:9]: Move to other file maybe?

File: src/modules/room/type/type.service.ts [hunk 16]
```diff
@@ -0,0 +1,78 @@
+import { prisma } from "../../../plugins/prisma";
+import {
+    CreateRoomTypeInput,
+    UpdateRoomTypeInput
+} from "./type.schema";
+import { id_not_found } from "./type.errors";
+
+export async function createRoomType(input: CreateRoomTypeInput) {
```
Comment (frederikpyt) [inline src/modules/room/type/type.service.ts:8]: Does this fail if the hotel with the given id does not exist?

File: src/modules/room/type/type.service.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    if (!RoomType) {
+        return id_not_found(id);
+    }
+
+    return RoomType;
+}
+
+export async function updateRoomType(input: UpdateRoomTypeInput) {
+    try {
+        const new_RoomType = await prisma.roomType.update({
```
Comment (frederikpyt) [inline src/modules/room/type/type.service.ts:37]: snake_CamelCase?

File: src/modules/room/type/type.errors.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
+export async function error_message(e: unknown) {
```
Comment (frederikpyt) [inline src/modules/room/type/type.errors.ts:1]: snake_case?

File: src/modules/room/type/type.errors.ts [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+    let message = String(e);
+
+    if (e instanceof Error) {
+        message = e.message;
+    }
+
+    return message;
+}
+
+export async function id_not_found(id: number) {
```
Comment (frederikpyt) [inline src/modules/room/type/type.errors.ts:11]: snake_case?

File: src/modules/room/type/type.controller.ts [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        return reply.badRequest(await error_message(e));
+    }
+}
+
+export async function browseRoomTypeHandler(
+    request: FastifyRequest,
+    reply: FastifyReply
+) {
+    try {
+        const RoomTypes = await request.redis.rememberJSON<RoomType[]>(CACHE_KEY_ROOM_TYPES, CACHE_TTL, async () => {
```
Comment (frederikpyt) [inline src/modules/room/type/type.controller.ts:46]: PascalCase?

File: src/modules/hotel/hotel.controller.ts [hunk 21]
```diff
@@ -6,20 +6,25 @@ import {
...
 } from "./hotel.service";
+import {
+    findRoomTypeByHotelId
+} from "../room/type/type.service";
 import {
     CreateHotelInput,
     UpdateHotelInput,
     ShowHotelParams,
-    DeleteHotelParams
+    DeleteHotelParams,
```
Comment (frederikpyt) [inline src/modules/hotel/hotel.controller.ts:16]: PascalCase?

File: src/modules/hotel/hotel.controller.ts [hunk 22]
```diff
@@ -92,11 +97,28 @@ export async function deleteHotelHandler(
...
+}
+
+export async function showHotelRoomTypeHandler(
+    request: FastifyRequest<{
+        Params: showHotelRoomTypeSchema;
+    }>,
+    reply: FastifyReply
+) {
+    try {
+        const room_types = await request.redis.rememberJSON<RoomType[]>(CACHE_KEY_HOTEL_ROOM_TYPES + request.params.id, CACHE_TTL, async () => {
```
Comment (frederikpyt) [inline src/modules/hotel/hotel.controller.ts:116]: snake_case

File: src/modules/hotel/hotel.schema.ts [hunk 23]
```diff
@@ -1,5 +1,6 @@
 import { z } from "zod";
 import { buildJsonSchemas } from "fastify-zod";
+import { RoomTypeSchema } from "../room/type/type.schema";
```
Comment (frederikpyt) [inline src/modules/hotel/hotel.schema.ts:3]: PascalCase?

File: src/modules/room/type/type.controller.ts [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+    browseRoomType,
+    showRoomType,
+    updateRoomType,
+    deleteRoomType
+} from "./type.service";
+import {
+    CreateRoomTypeInput,
+    UpdateRoomTypeInput,
+    ShowRoomTypeParams,
+    DeleteRoomTypeParams
```
Comment (frederikpyt) [inline src/modules/room/type/type.controller.ts:13]: PascalCase?

File: src/modules/room/type/type.controller.ts [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+export async function showRoomTypeHandler(
+    request: FastifyRequest<{
+        Params: ShowRoomTypeParams;
+    }>,
+    reply: FastifyReply
+) {
+    try {
+        const RoomType = await request.redis.rememberJSON<RoomType>(CACHE_KEY_ROOM_TYPE+request.params.id, CACHE_TTL, async () => {
```
Comment (frederikpyt) [inline src/modules/room/type/type.controller.ts:63]: PascalCase?

File: src/modules/room/type/type.controller.ts [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+
+export async function updateRoomTypeHandler(
+    request: FastifyRequest<{
+        Body: UpdateRoomTypeInput;
+    }>,
+    reply: FastifyReply
+) {
+    try {
+        await request.redis.invalidateCaches(CACHE_KEY_ROOM_TYPE + request.body.id, CACHE_KEY_ROOM_TYPES, CACHE_KEY_HOTEL_ROOM_TYPES);
+        const RoomType = await updateRoomType(request.body);
```
Comment (frederikpyt) [inline src/modules/room/type/type.controller.ts:81]: PascalCase?

File: src/modules/room/type/type.controller.ts [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+
+export async function deleteRoomTypeHandler(
+    request: FastifyRequest<{
+        Params: DeleteRoomTypeParams;
+    }>,
+    reply: FastifyReply
+) {
+    try {
+        await request.redis.invalidateCaches(CACHE_KEY_ROOM_TYPE + request.params.id, CACHE_KEY_ROOM_TYPES, CACHE_KEY_HOTEL_ROOM_TYPES);
+        const RoomType = await deleteRoomType(Number(request.params.id));
```
Comment (frederikpyt) [inline src/modules/room/type/type.controller.ts:97]: PascalCase?

**Review by frederikpyt — CHANGES_REQUESTED** [review]
Multiple different variable naming conventions

File: src/modules/room/type/__test__/createRoomType.test.ts (hunk 2 above)
Comment (kristianbinau) [reply src/modules/room/type/__test__/createRoomType.test.ts:139]: 💀

File: src/modules/room/type/__test__/createRoomType.test.ts (hunk 3 above)
Comment (kristianbinau) [reply src/modules/room/type/__test__/createRoomType.test.ts:14]: fuuuuu

File: src/modules/hotel/__test__/deleteHotel.test.ts [hunk 28]
```diff
@@ -43,6 +44,38 @@ describe("DELETE /api/hotel/:id", () => {
         expect(count).toBe(0);
     });
 
+    it("should return status 204 and delete a hotel", async () => {
```
Comment (frederikpyt) [inline src/modules/hotel/__test__/deleteHotel.test.ts:47]: More descriptive

File: src/modules/room/type/type.service.ts [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+import {
+    CreateRoomTypeInput,
+    UpdateRoomTypeInput
+} from "./type.schema";
+import { showHotel } from "../../hotel/hotel.service";
+import { idNotFound } from "./type.errors";
+
+export async function createRoomType(input: CreateRoomTypeInput) {
+    // Only using it's thrown error.
+    await showHotel(input.hotel_id);
```
Comment (frederikpyt) [inline src/modules/room/type/type.service.ts:11]:
```suggestion
    const roomType = await findRoomTypeById(id);
    if (!roomType) {
        return idNotFound(id);
    }
```

File: src/modules/room/type/type.service.ts [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+    if (!roomType) {
+        return idNotFound(id);
+    }
+
+    return roomType;
+}
+
+export async function updateRoomType(input: UpdateRoomTypeInput) {
+    // Only using it's thrown error.
+    await showHotel(input.hotel_id);
```
Comment (frederikpyt) [inline src/modules/room/type/type.service.ts:41]:
```suggestion
    const roomType = await findRoomTypeById(id);
    if (!roomType) {
        return idNotFound(id);
    }
```

File: prisma/schema.prisma [hunk 31]
```diff
@@ -18,14 +18,29 @@ model User {
...
+
+model RoomType {
     id          Int      @id @default(autoincrement())
     name        String   @db.VarChar(255)
     description String?  @db.Text
-    address     String   @db.VarChar(255)
+    size        String?  @db.VarChar(255)
+    price       Float    @db.Double
+    hotel       Hotel    @relation(fields: [hotel_id], references: [id], onDelete: Cascade)
+    hotel_id    Int
```
Comment (frederikpyt) [inline prisma/schema.prisma:41]: snake?

**Review by frederikpyt — CHANGES_REQUESTED** [review]

**Review by frederikpyt — APPROVED** [review]
🥇

**Pull request merged**